[Optional feature]: Staticman Nested Comment Support

VincentTam opened the pull request:
# Motivation
Provide a *free* commenting service.

Hugo's built-in Disqus has no Markdown support.  [Staticman](https://staticman.net) is open-source and transmits HTML form data as YML files to be merged in GitHub/GitLab repos.  Despite its MIT license, the project had been bundled with GitHub for years.  Thanks to Nicholas Tsim's PR, Staticman now supports GitLab.

To add Staticman, I've combined Minimal Mistakes and Beautiful Hugo's code so as to avoid breaking the original modals.  I am *not* good at CSS at the moment, so the visual layout will be strange.  It took me hours to understand the role of `container`.  During local preview, the SASS comes out.  However, when it's deployed to GitLab, this has disappeared. However, the logic is already there at https://vincenttam.gitlab.io/introduction.  Web developers should be able to fix the display much more efficiently than I do.  I'm creating this PR to publish my code.

![screenshot_2018-12-18-17-48-55](https://user-images.githubusercontent.com/5748535/50194630-11056c80-033c-11e9-8145-50ff0e1d3c2a.png)


# Features

- [X] Staticman comments
- [X] Reply function
- [X] sample config files (site, repo, GitLab config) in `exampleSite`
- [x] CSS
- [x] Documentation

# Quick guide to test this PR
I assume the user has installed his/her Hugo theme(s) as a Git submodule.
```sh
$ git remote -v
upstream	https://github.com/vickylai/hugo-theme-introduction.git (fetch)
upstream	https://github.com/vickylai/hugo-theme-introduction.git (push)
... # other remote omitted
$ git checkout -b pr119 # test on a new branch pr119
$ git fetch upstream pull/119/head # git pull will be rejected
$ git merge FETCH_HEAD # manually merge the this PR against branch pr119
$ cd <your-blog>
... # edit your .gitmodules with url="<repo-containing-pr119>" and branch = "pr119"
$ git submodule sync # inform Git the changes in .gitmodules
$ git submodule update --remote --recursive # switch to the HEAD of your cloned repo for the theme
```

# On the recent Staticman's public API bottleneck

Due to eduardoboucas/staticman#227, the parameter `api` in `config.toml` defaults to the API address associated with @staticmanlab.

hanzei commented:
@VincentTam First of all: Thank you very much for the time and effort you put into this! The code you wrote it very elegant and fits well into hugo.

But I'm not sure if this is something I want to add to this theme. The theme is about setting up a simple and minimalistic personal site. Blogging is just a site feature of this theme. I assume most people just want to set up there site and don't care about comments.

@VincentTam Thoughts on this? Out of interests: Why did you put work into this theme and not any other?

@vickylai Would love to hear your thoughts on this!

VincentTam commented:
Thank you for your reply despite my unfinished work.  Since I'm not a developer, I publish it here for interested users.

This PR aims at providing comment support alternative to _existing_ Disqus support.  For Disqus vs Staticman, you may google for the rationale of preferring the later.  (I'm replying on mobile.)

Whether comment is important is subjective.  A theme designer may (or may not) offer the freedom for its users to enable this feature.  With Staticman's config parameters missing or blank (default), such features will be disabled.

Looking back, my code _isn't_ the best.  I've messed up the code for the reply target in each comment reply.  Please see the last point for details.

I _did_ integrate Staticman into _other_ templates.  You may consult

1. GitLab Pages/Hugo MR ?
2. GitLab Pages/Jekyll MR ?
3. https://framagit.org/staticman-gitlab-pages

In fact, this is a derivative of halogenica/beautifulhugo#222, which aims to fix the logic problem mentioned in https://github.com/dancwilliams/networkhobo/issues/373.

---

IMHO, interactions with readers are beneficial to site owners, who often get great ideas from readers.  Moreover, having static comments (with Gravatar of other users) is great in terms of SEO because search engines view this as the activities of the page.

vickylai commented:
> But I'm not sure if this is something I want to add to this theme. The theme is about setting up a simple and minimalistic personal site. Blogging is just a site feature of this theme. I assume most people just want to set up there site and don't care about comments.
> 
> @vickylai Would love to hear your thoughts on this!

I’d leave this one to the community to decide, if anyone wishes to weigh in. Introduction is meant to be a minimalist personal site, but I’m not opposed to the addition of comment functionality, as long as it is clearly optional. Especially with Staticman over Disqus.

I would like to be able to take a look at the styling within the next couple weeks. If anyone else wants to give it a go in the meantime, please do!

VincentTam commented:
I've recently revised the template code in [Huginn](https://framagit.org/staticman-gitlab-pages/huginn/compare/2288c39...4a9a46d#a5ecb92e0a1d382f4ff6d8fd012aaf69519aa0ac).  The reasons are two-folded.

1. Unnecessary JS code for smooth-scrolling and complicated Go-HTML template code

    That was unintentionally copied from the original PR for Beautiful Hugo, which is the theme that I'm now using and tweaking, and which has _existing_ Staticman integration code that I _don't_ wish to break.

2. Mixed JS code and Go template code for internationalization.

    I copied this from the Jekyll theme Minimal Mistakes.  A more unobstrusive approach would be to load site data into Go-HTML template files (with some hidden elements), and operated on them using JavaScripts in the assets, instead of injecting the locale data strings into the JS code.

victoriadotdev commented:
Hello @VincentTam! I made some updates to your work but am unable to push to this PR. If you'd like to keep it going, would you kindly follow [these instructions](https://help.github.com/en/articles/allowing-changes-to-a-pull-request-branch-created-from-a-fork#enabling-repository-maintainer-permissions-on-existing-pull-requests) to grant me permission to make edits?

If you don't want to do that, you can still look over my changes in the new [branch](https://github.com/victoriadotdev/hugo-theme-introduction/tree/VincentTam-staticman3) and see if there's anything you'd like to implement on your own.

VincentTam commented:
@victoriadotdev Thanks for merging `staticman3` and your changes.  Regarding the permissions problem, that's unexpected.  I usually follow the defaults for PR submission and leave the checkbox "allow changes from maintainers" checked.  As a result, that's what I can see from my qutebrowser on Xubuntu 18.04.

![Screenshot_2019-07-22_08-02-08](https://user-images.githubusercontent.com/5748535/61609810-9563f900-ac57-11e9-82ab-523dfc426db7.png)

At the bottom right corner of the screenshot, I can see that the relevant checkbox is checked.  I'm going to contact GitHub support for that.

VincentTam commented:
@victoriadotdev I've just got a reply from GitHub, who suggested me to uncheck and check the checkbox.  I've done that, and please verify whether the write permission problem has been solved.

VincentTam commented:
> Hello @VincentTam! I made some updates to your work but am unable to push to this PR. If you'd like to keep it going, would you kindly follow [these instructions](https://help.github.com/en/articles/allowing-changes-to-a-pull-request-branch-created-from-a-fork#enabling-repository-maintainer-permissions-on-existing-pull-requests) to grant me permission to make edits?

@victoriadrake Another update: Despite further discussions with GitHub Support staff, she couldn't find out the reason why you were unable to push to this PR.  Since this PR's submission, the write permissions from maintainers have always been enabled.  I tried asking if your recent change of user name had contributed to this error, by she was unable to verify this theory after some tests.

It seems that this push failure is due to past conflicts of translation files, in which your user name appeared at the bottom.

1. I tried to port a section of UI text from Minimal Mistakes and Beautiful Hugo 6 months agos.
2. You've recently changed your user name, which appeared at the bottom of the file.

As a result, each translation file is modified on both sides at the end of the file, and this causes a merge conflict.  As can be seen from the commit tree of the HEAD of this PR, I've merged your changes this week, so the conflicts have been resolved.  Hope this solves your difficulties in committing against the head of this PR.

* * *

# A self-review on the recent changes of this PR

1. Replaced Beautiful Hugo's logic for looping over `.Site.Data.comments` with my own one: the former sets the `id` attribute of each static comment as `comment-N`.  I've changed that to the `_id` field of the YML data file due to this scenario: in case of `moderation:false`, comments are published once submitted.  Site owners might want to delete a few unwanted ones afterwards.  After such deletion, the URL of every static comment after the deleted comment would change.  Existing links pointing to those comments would not function as desired.  Those such possibility is marginal for minimalistic and personal static sites, but this issue of URL shifting can be avoided by using the `_id` field unique to each static comment.

    As a consequence, I've removed the JS code for smooth scroll in the Staticman script.

2. Moved Staticman JS script from a `<script>` tag under `layouts/partials/js` to the assets folder.  Now, the role of `layouts/partials/js/staticman.html` is to
    - include a `<script>` tag pointing to `assets/js/staticman.js`
    - excecute the Hugo code inside `assets/js/staticman.js` for constructing `<form action>` URL with JavaScript due to the behavior of some search bots described in [this blog post](https://www.willmaster.com/library/manage-forms/bot-block-for-forms.php).  An example of form action URL in JavaScript: [this blog post](https://team-parallax.com/use-selfhosted-staticman-with-gitlab-1).

    ![Introduction + Staticman demo site new UI](https://framapiaf.s3.framasoft.org/framapiaf/media_attachments/files/004/298/254/original/9743843b174cf891.png)

3. Improved JS code for setting reply target when the reply button is clicked.
https://github.com/victoriadrake/hugo-theme-introduction/blob/8b1e782b443c8f0e9b88208f793beb41dddb17e1/layouts/partials/staticman.html#L38
https://github.com/victoriadrake/hugo-theme-introduction/blob/8b1e782b443c8f0e9b88208f793beb41dddb17e1/layouts/partials/staticman.html#L63
The `title='{{ ... }}'` here is crucial in the construction of internal referencing of nested comments.  That would lead to an unhelpful tooltip text showing the `_id` of the static comment's eldest ancestor.
    - A CSS rule can be added to hide this if you want; or
    - store this piece of info elsewhere (say, in the form of a hidden `<span>` tag) under the `<article>` tag for the static comment.
4. Added a functionality for displaying replying target name and avatar on top of the comment form when the "reply to this comment" button is clicked.  Clicking a :x: button on the right of the reply target display would remove the reply target.  Clicking on another reply button under another comment would update the reply target display.
5. Improved display and introduced three SASS variables in Staticman SASS file for calculating the appropriate left margin of a comment reply.
https://github.com/victoriadrake/hugo-theme-introduction/blob/8b1e782b443c8f0e9b88208f793beb41dddb17e1/assets/sass/_staticman.sass#L73

VincentTam commented on assets/js/staticman.js:
```diff
@@ -0,0 +1,100 @@
+  $('.page__comments .comment').on('click', '.reply-btn', function (evt){
+    resetReplyTarget();
+    var cmt = $(evt.delegateTarget);
+    $('.page__comments input[name="fields[replyThread]"]').val(this.title);
```
If you find my approach of storing the `_id` of a comment's eldest ancestor in the `title` attribute of an `<a>` tag not satisfying, I can change the `this.title` to `replyThread`, where `var replyThread = cmt.find('.someClassName')`.

VincentTam commented on README.md:
```diff
@@ -158,6 +158,89 @@ Set `googleAnalytics` in `config.toml` to activate Hugo's [internal Google Analy
+1. Add Staticman bot/app, depending on your Git service provider.
+    - GitHub: choose either one of the following method
+        + GitHub App: refer to issue
+        [https://github.com/eduardoboucas/staticman/issues/243](https://github.com/eduardoboucas/staticman/issues/243)
```
@staticmanapp and `https://api.staticman.net/v2` no logner work, as reported in eduardoboucas/staticman#306.

VincentTam commented on README.md:
```diff
@@ -158,6 +158,89 @@ Set `googleAnalytics` in `config.toml` to activate Hugo's [internal Google Analy
+            Then help **[@staticmanlab](https://github.com/staticmanlab)**
+            accept the invitation by going to
+
+                https://staticman3.herokuapp.com/v3/connect/github/<username>/<repo-name>
```
[Staticman's documentation](https://staticman.net) only covers v2, so there's a need to give a self-contained setup guide.

VincentTam commented on assets/js/staticman.js:
```diff
@@ -0,0 +1,100 @@
+
+    $.ajax({
+      type: $(this).attr('method'),
+      url: [endpoint, 'v3/entry', gitProvider, repo, branch, 'comments'].join('/'),
```
Separated POST request URL in JS code to prevent abuses by spam bots through scraping the generated HTML source code.

VincentTam commented on assets/sass/_staticman.sass:
```diff
@@ -0,0 +1,95 @@
+  /* Static comments */
+  $avatar__size: 5rem
+  $avatar__sep: 3rem
+  $reply__indent: 2rem
```
Feel free to change the values here. Since I'm not a web designer, I haven't considered the mobile loading issue.
